Catch a possible exception when trying to resolve from HKLM.

directhex opened the pull request:
On Mono, RegistryKey.OpenSubKey will attempt to create the registry hive's on-disk folder, if it doesn't exist. For HKLM, that folder is in /etc/mono/registry - but only `root` has write access to /etc, so the constructor throws an exception. We can just try{}catch{} this, since we can't risk calling the constructor, and if it throws then it doesn't exist (so obviously there's nothing of value inside).

radical commented:
Maybe we should be calling [this](https://github.com/directhex/msbuild/blob/df3d5d16a751c27739d5fb2d81fbb848aa031d61/src/XMakeTasks/AssemblyFolder.cs#L123-L133) only on Windows.

rainersigwald commented:
If Mono doesn't store assembly-folder information there that sound entirely reasonable @radical.

radical commented:
I think we should avoid depending on assembly folders from registry here. We any way, have the new assembly folders from file feature now, and we could utilize that if we need to add more "default" search paths.

directhex commented:
@radical I agree, a Windows check further up the chain makes more sense.

Commit:
Catch a possible exception when trying to resolve from HKLM

On Mono, RegistryKey.OpenSubKey will attempt to create the registry hive's on-disk folder, if it doesn't exist. For HKLM, that folder is in /etc/mono/registry - but only root has write access to /etc, so the constructor throws an exception. Let's avoid looking in the registry at all, on non-Windows.

directhex commented:
Better?

rainersigwald commented on src/XMakeTasks/AssemblyFolder.cs:
```diff
@@ -113,18 +113,21 @@ private static void CreateAssemblyFolders()
-                @"SOFTWARE\Microsoft\.NETFramework\AssemblyFolders",
-                s_assemblyFolders
-            );
+            if (!(Environment.OSVersion.Platform == System.PlatformID.MacOSX || Environment.OSVersion.Platform == System.PlatformID.Unix))
```
Can you use `if (NativeMethodsShared.IsWindows)`?

rainersigwald reviewed (commented):
LGTM with a tweak to match other similar checks.

directhex commented:
Changed.

radical commented:
LGTM!
